servicelayerfunctionalities.cc

```diff
@@ -0,0 +1,328 @@
+    fromget = clientKey.get("reply"+currentChirp.id());
+    if(fromget.size() == 0) {
+      continue;
+    }
+
+    chirp::ChirpReplies chirpReplies = convertToChirpReplies(fromget[0]);
+    for (int i = 0; i < chirpReplies.chirps_size(); i++) {
+      myqueue.push(chirpReplies.chirps(i));
+      chirpToSend.push_back(chirpReplies.chirps(i));
+    }
+  }
+  //Sort by timestamp
+  std::sort(chirpToSend.begin(), chirpToSend.end(), [](chirp::Chirp &a, chirp::Chirp &b){
+    return a.timestamp().useconds() < b.timestamp().useconds();
+  });
+  for (const auto &i :chirpToSend) {
+    if (i.id() == mainChirp->id()) {
+      continue;
+    }
+    chirp::Chirp* c = response->add_chirps();
+    copyChirp(c, i);
+  }
+  return Status::OK;
+}
+Status ServerForCommandLineClient::monitor(ServerContext* context, const MonitorRequest* request, 
+  ::grpc::ServerWriter< ::chirp::MonitorReply>* writer) {
+  //Get Current TimeStamp
+  std::time_t seconds = std::time(nullptr);
+  int64_t seconds_since_epoch= static_cast<int64_t>(seconds);
+  int64_t microseconds_since_epoch = std::chrono::duration_cast<std::chrono::microseconds>(std::chrono::system_clock::now().time_since_epoch()).count();
+  
+  ClientForKeyValueStore clientKey(grpc::CreateChannel("localhost:50000", grpc::InsecureChannelCredentials()));
+  
+  auto fromget = clientKey.get(request->username());
+  if (fromget.size() == 0) {
+    return Status::CANCELLED;
+  }
+
+  std::set<std::string> chirpsent;
+  chirp::User user; //Main User
+  user.ParseFromString(fromget[0]);
+
+  chirp::MonitorReply reply;
+  chirp::Followers followers = user.followers();
+  /*
+    Continuously look through all user's followers, and their chirps. Keep all sent chirps in a set called
+    chirpsent. Keep looking for new chirps with this while loop.
+  */
+  while (true) {
+    if (context->IsCancelled()) {
+      break;
+    }
+    for (int i = 0; i < followers.username_size(); i++) {
+      auto allfollowers = clientKey.get(followers.username(i));
+
+      if(allfollowers.size() > 0) {
+        chirp::User userFollowers;
+        std::string temp = allfollowers[0];
+        userFollowers.ParseFromString(temp);
+
+        for(int j = 0; j < userFollowers.chirps_size(); j++) {
+          auto it = chirpsent.find(userFollowers.chirps(j).id());
+
+          if (it == chirpsent.end()) {
+            if (userFollowers.chirps(j).timestamp().useconds() > microseconds_since_epoch) {
+              chirpsent.insert(userFollowers.chirps(j).id());
+
+              chirp::Chirp* c = reply.mutable_chirp();
+              copyChirp(c, userFollowers.chirps(j));
+              writer->Write(reply);
+              reply.clear_chirp();
+            }
+          
+          } 
+        }
+      }
+    }
+    std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+  }
+  return Status::OK;
+}
+void ServerForCommandLineClient::copyChirp(chirp::Chirp* c, const chirp::Chirp &r) {
```

Comment:
Variable names should be more descriptive

Reply:
done